### ISSUE TITLE
Updated to latest version - labels replaced by question-marks

### ISSUE THREAD

**Issue by ghost** [bvaughn/angular-form-for#81]
Hi Brian

I've been trying to update formFor to the latest version (so that I can find the disable attribute to a field), but have encountered a problem that I'm having trouble solving.  All my labels have been replaced by question-marks as shown below:

![image](https://cloud.githubusercontent.com/assets/5274895/5265939/21be42b8-7a92-11e4-9dc2-c3370ef8d718.png)

Have you got any idea why that is happening?

Regards
Hans

**Comment by bvaughn** [comment]
Hi @EnterpriseDT,

I hope you don't mind my reaching out to you in this way, but I have a small request I'd like to make of you. (And there will be no hard feelings if you don't have the time!)

I've recently started created a new JavaScript library named Task Runner (github.com/bvaughn/task-runner). If you're willing and able to take a few moments to look over the library, I'd love to hear your feedback. It's in the early stages (only 1 release) so there's still a lot of room for changes.

I think the library has a lot of potential, but before I push it out via Reddit and other forums, I'd love to get some unbiased feedback.

Cheers,
Brian

**Comment by EnterpriseDT** [comment]
Hi Brian

I'll have a proper look at it soon, but reading the first section I already
have some feedback that I think could be helpful to you.

The first paragraph should state as concisely as possible what the
framework does in order to grab the developer's attention.  The current
paragraph says 1) it makes JS development easier (which tells me nothing
since you could say that of any library) and 2) that it doesn't depend on
other libraries (is this really the second-most important thing about the
library?!).

In the second paragraph you introduce tasks: "A task is kind of like a
JavaScript Promise but more powerful."  This is a pretty vague - can you be
a bit more specific?  You then mention "task-based applications", which is
a term with which I'm not familiar.  I googled "task-based application"
<https://www.google.com?q=%22task-based+application%22> and it just came up
with a bunch of fairly old academic papers, so it doesn't seem to be a term
that's popularly used.

So at this point I'm wondering if what you call a "task" is something that
you've come up with or a Javascript implementation of a commonly used
design-pattern.  If it's the latter then a reference to something that
defines it might be useful.  If it's the former then it's even more
important that you find a way to describe it really concisely.

A lot of readers probably won't read any more than this, since the
following section looks like it's starting to delve into details and they
probably aren't yet 'sold'.
​
This might seem annoying and pedantic since you're no doubt more interested
in the feedback on the technology, but I think it's important.

I'll get back to you with more techy feedback soon.

Hans

**Comment by bvaughn** [comment]
This is fantastic feedback already, Hans. Thank you for taking the time to share it!

I've updated the main README to (hopefully) address what you're talking about. Rather than making vague claims, I just jump directly into examples/details. Hopefully this reads better!

Looking forward to hearing what else you think about the library. :)

**Comment by bvaughn** [comment]
FWIW your comment also inspired me to spent a big chunk of time overhauling the API docs (bvaughn.github.io/task-runner/#/documentation/index) this weekend for better readability and lots of use-case examples. I hope this further addresses your concerns.

**Comment by EnterpriseDT** [comment]
Hi Brian

The new documentation layout is great.  I really like the slide style and
the first page "What is a Task?" cuts straight to the chase.

The front-page <http://bvaughn.github.io/task-runner/#/> still has the
vague sentence that used to be at the start of the documentation, i.e.
"Task Runner is a collection of low-level libraries designed to make
JavaScript application development easier.", but I guess you've decided to
leave that as is, so perhaps I shouldn't repeat that point.

The use of 'tr' as the name for the reference to TaskRunner stuck out when
I first saw it.  Firstly, somehow it looks like a local variable to me so I
looked above to see if it was declared anywhere.  Secondly, for me, 'tr' is
the <tr> element, so I had to think a bit before realizing that it stands
for TaskRunner.  Is there anything wrong with just using 'taskRunner'?  Or
maybe I'm just a bit slow.

Anyway, for me, the first this-is-cool moment was on the 'Chaining tasks'
page where you wrote "1. First Task A and B will be run (in parallel). 2.
If they succeed Task C will be run next."  This is something that's not
trivial to implement, so this is cool.  Also the way that you can embed
logic in the chain is neat.  One thing - why do you use 'or' for the
error-handler?

I still feel like you could get to the this-is-cool point a bit quicker.
Why not start with a 'hello world' type example, instead of explaining the
details of creating tasks.  For the initial example, I'd just show the
simplest way of creating tasks and then show them being chained.  Each task
could just be a setTimeout 'task' that would log something when it's
finished.  You could then include the log output after the chain has
completed.

I hope I'm not being too much of a crank.  In any case, please don't feel
obliged to do anything about my feedback, since you already have really.

More next time.  :)


Hans


On 18 February 2015 at 08:10, Brian Vaughn <notifications@github.com> wrote:

> FWIW your comment also inspired me to spent a big chunk of time
> overhauling the API docs (
> bvaughn.github.io/task-runner/#/documentation/index) this weekend for
> better readability and lots of use-case examples. I hope this further
> addresses your concerns.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bvaughn/angular-form-for/issues/81#issuecomment-74765568>
> .
>



-- 
Regards
Hans Andersen
EnterpriseDT

**Comment by bvaughn** [comment]
Thanks Hans! This is great feedback.

The choice of "or" was just because it felt like a decent word to me. (I also aliased "else" and "otherwise" methods that point to the "or" method). But if you have a better suggestion, I'm open to it. Naming things can be hard. :D

I like the suggestion for an example on the intro page! I'll add that this evening after work.

Thanks again.

**Comment by EnterpriseDT** [comment]
>
> The choice of "or" was just because it felt like a decent word to me. (I
> also aliased "else" and "otherwise" methods that point to the "or" method).
> But if you have a better suggestion, I'm open to it. Naming things can be
> hard. :D
>
But isn't it only called when there's an error?  In that case I'd expect it
to be called something like 'onError' or 'error'.  The words, 'or', 'else'
and 'otherwise' imply that there's a condition somewhere (as in an
if-statement), which there isn't really is there?

> I like the suggestion for an example on the intro page! I'll add that this
> evening after work.
>
Excellent.  I look forward to seeing it.

Hans

**Comment by bvaughn** [comment]
It is called only when there's an error in the previous chain-link. I avoided using a name like "errored" though because I thought it might be confused with the entire chain erroring, which only happens if no `or()` is specified (or if all of the `or` fallbacks fail).

I added a little bit of a teaser example to the intro page, based on your feedback. Hopefully it's enough to get people interested so they keep reading.

**Comment by EnterpriseDT** [comment]
That's a great way to communicate the benefits.  In fact, I feel I
understand it better already just by looking at that.

You seem to imply that a task can include complex user interactions, such
as authenticating a user.  Is that really true?

Can you use it to handle the state-transitions in an Angular app that uses
UI-Router?

Hans


On 24 February 2015 at 16:24, Brian Vaughn <notifications@github.com> wrote:

> It is called only when there's an error in the previous chain-link. I
> avoided using a name like "errored" though because I thought it might be
> confused with the entire chain erroring, which only happens if no or() is
> specified (or if all of the or fallbacks fail).
>
> I added a little bit of a teaser example to the intro page, based on your
> feedback. Hopefully it's enough to get people interested so they keep
> reading.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bvaughn/angular-form-for/issues/81#issuecomment-75706090>
> .
>



-- 
Regards
Hans Andersen
EnterpriseDT

**Comment by bvaughn** [comment]
Ooh, good! I'm glad to hear.

Yes! Tasks can include complex interactions and yes, you totally can use it for state-transitions for UI-Router or other similar frameworks. :) Check out the [Promise task](http://bvaughn.github.io/task-runner/#/documentation/tr.Promise) for an example of how a task can be created from a promise or converted to a promise.

I also recently added the [Resolver task](http://bvaughn.github.io/task-runner/#/documentation/tr.Resolver) when thinking of how to handle more complicated things like state-transitions, where you have a number of dependencies (like UI Router's resolve block) that may fail. The resolver lets you specify fall-back behavior.

All of this (the framework and my ideas about it) are new though. I could be convinced to change. :D

**Comment by bvaughn** [comment]
PS. For a complex task like authenticating a user, I imagine you could use something like [Composite task](http://bvaughn.github.io/task-runner/#/documentation/tr.Composite) [Graph task](http://bvaughn.github.io/task-runner/#/documentation/tr.Graph) to wrap up a couple of lower-level tasks:

1. Perhaps a [Listener task](http://bvaughn.github.io/task-runner/#/documentation/tr.Listener) that listens to for a form-submit event.
2. Then perhaps an [Xhr task]((http://bvaughn.github.io/task-runner/#/documentation/tr.Xhr) that submits authentication details to a server. Alternately you could use the [Promise task]((http://bvaughn.github.io/task-runner/#/documentation/tr.Promise) to wrap a jQuery or Angular GET/POST request.

The nice thing about this is that you can keep the parts of your authentication modular and focused, while at the same time treating the composite/graph of them as a simple task (since that's all it is). And Task Runner manages chaining them together, interrupting, error-handling, etc.

**Comment by EnterpriseDT** [comment]
This sounds really useful.  I had a fair bit of trouble implementing the
state-transitions in my app.

E.g. we have different ways of creating user accounts.  If the new user is
just trialing our product then we only ask for a user-name, password and an
e-mail address, but if they're in the process of ordering something then we
need a lot more details (billing contact, company address, etc).  For this
reason we separated out the pages required for entering a user-name,
password and e-mail address into a generic 'user' module and placed the
ordering stuff in our 'shop' module.  It was then quite tricky to get the
state-transitions to work correctly for the different scenarios.

Intuitively it sounds like we'd be able to implement the generic part of
the task as one (child) task and then call that from the more complex task
in the ordering module.  Which would be very cool!

Unfortunately, I'm not really in a position to be able to try that at the
moment since I've moved on to another task.

On 24 February 2015 at 16:44, Brian Vaughn <notifications@github.com> wrote:

> PS. For a complex task like authenticating a user, I imagine you could use
> something like Composite task
> <http://bvaughn.github.io/task-runner/#/documentation/tr.Composite> Graph
> task <http://bvaughn.github.io/task-runner/#/documentation/tr.Graph> to
> wrap up a couple of lower-level tasks:
>
>    1. Perhaps a Listener task
>    <http://bvaughn.github.io/task-runner/#/documentation/tr.Listener>
>    that listens to for a form-submit event.
>    2. Then perhaps an [Xhr task]((
>    http://bvaughn.github.io/task-runner/#/documentation/tr.Xhr) that
>    submits authentication details to a server. Alternately you could use the
>    [Promise task]((
>    http://bvaughn.github.io/task-runner/#/documentation/tr.Promise) to
>    wrap a jQuery or Angular GET/POST request.
>
> The nice thing about this is that you can keep the parts of your
> authentication modular and focused, while at the same time treating the
> composite/graph of them as a simple task (since that's all it is). And Task
> Runner manages chaining them together, interrupting, error-handling, etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bvaughn/angular-form-for/issues/81#issuecomment-75707595>
> .
>



-- 
Regards
Hans Andersen
EnterpriseDT

**Comment by bvaughn** [comment]
Understandable.

It's encouraging to hear that you seem to think Task Runner offers a potential solution to that problem. Hopefully others will think the same, if they stumble upon the project. :)
